### PR TITLE
Operator Clustering support

### LINKED ISSUE #9730: Clustering support in operator
Make sure Keycloak can properly (manually) horizontally scale.

### PR DESCRIPTION
**Pull request by andreaTP**
Resolves #9730 
Take over #10249 

@vmuzikar I believe that I addressed the review comments in the other PR, this should be ready for the next round of review.

### CODE REVIEW

**Review by abstractj — COMMENTED** [review]
@andreaTP could you please rebase?

**Comment by andreaTP** [comment]
@abstractj done

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.util.Optional;
+
+public class KeycloakDiscoveryService extends OperatorManagedResource implements StatusUpdater<KeycloakStatusBuilder> {
+
+    private Service existingService;
+
+    public KeycloakDiscoveryService(KubernetesClient client, Keycloak keycloakCR) {
+        super(client, keycloakCR);
+        fetchExistingService().ifPresent(s -> this.existingService = s);
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java:40]: Just a nit pick, not a blocker, but this is IMHO a great example of overengineering. 🙂 What if we could do something like this:
```java
existingService = fetchExistingService(); // this can be null
```
Well, isn't it simpler? 😉

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    protected Optional<HasMetadata> getReconciledResource() {
+        return Optional.of(
+                Optional.ofNullable(existingService)
+                    .map(a -> {
+                            a.setSpec(getServiceSpec());
+                            return a;
+                            })
+                .orElse(newService()));
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java:61]: Again, feels a bit overengineered and hard to read. I know we had something similar in the [original PR]() but this is even more complicated.

But just a nit pick, not a blocker.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.util.Optional;
+
+public class KeycloakService extends OperatorManagedResource implements StatusUpdater<KeycloakStatusBuilder> {
+
+    private Service existingService;
+
+    public KeycloakService(KubernetesClient client, Keycloak keycloakCR) {
+        super(client, keycloakCR);
+        fetchExistingService().ifPresent(s -> this.existingService = s);
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:40]: Same comment about overengineering as in the discovery service.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                .build();
+        return service;
+    }
+
+    private Optional<Service> fetchExistingService() {
+        return Optional.ofNullable(client
+                .services()
+                .inNamespace(getNamespace())
+                .withName(getName())
+                .get());
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:79]: Same comment about overengineering as in the discovery service.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    public void updateStatus(KeycloakStatusBuilder status) {
+        if (existingService == null) {
+            status.addNotReadyMessage("No existing Discovery Service found, waiting for creating a new one");
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:84]:
```suggestion
            status.addNotReadyMessage("No existing Keycloak Service found, waiting for creating a new one");
```

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                .atMost(2, MINUTES)
+                .pollDelay(5, SECONDS)
+                .ignoreExceptions()
+                .untilAsserted(() -> {
+                    var conditions = crSelector
+                            .get()
+                            .getStatus()
+                            .getConditions();
+
+                    assertThat(getCondition(conditions, READY).getStatus()).isFalse();
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:63]: Please use:
https://github.com/keycloak/keycloak/blob/55315c18b146086fe2e0d1b290dd0106f5206ada/operator/src/test/java/org/keycloak/operator/utils/CRAssert.java#L28

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                .atMost(2, MINUTES)
+                .pollDelay(5, SECONDS)
+                .ignoreExceptions()
+                .untilAsserted(() -> {
+                    var conditions = crSelector
+                            .get()
+                            .getStatus()
+                            .getConditions();
+
+                    assertThat(getCondition(conditions, READY).getStatus()).isTrue();
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:86]: Please use:
https://github.com/keycloak/keycloak/blob/55315c18b146086fe2e0d1b290dd0106f5206ada/operator/src/test/java/org/keycloak/operator/utils/CRAssert.java#L28

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                });
+
+        // get the service
+        var service = new KeycloakService(k8sclient, kc);
+        String url = "http://" + service.getName() + "." + namespace + ":" + KEYCLOAK_DISCOVERY_SERVICE_PORT;
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:91]:
```suggestion
        String url = "http://" + service.getName() + "." + namespace + ":" + KEYCLOAK_SERVICE_PORT;
```

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.List;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.keycloak.operator.Constants.KEYCLOAK_DISCOVERY_SERVICE_PORT;
+import static org.keycloak.operator.v2alpha1.crds.KeycloakStatusCondition.READY;
+
+@QuarkusTest
+public class ClusteringE2EIT extends ClusterOperatorTest {
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:23]: We're missing any test for the discovery service. This should be ideally an Infinispan test that would check the nodes really see each other, but I'm not sure how to do that and if it's even possible. Maybe @hmlnarik might have some tips?

Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:23]: We should also test the durability of the services. That specs are reconciled if modified externally.

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java [hunk 10]
```diff
@@ -67,4 +73,39 @@ public static void waitForKeycloakToBeReady(KubernetesClient client, Keycloak kc
                     CRAssert.assertKeycloakStatusCondition(currentKc, KeycloakStatusCondition.HAS_ERRORS, false);
                 });
     }
+
+    public static String inClusterCurl(KubernetesClient k8sclient, String namespace, String url) {
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:77]: I like it that it's generic now. 👍

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@andreaTP Thanks for overtaking the PR. Added a few comments.

Please also update the commit message with [`Co-authored-by`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java (hunk 1 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java]: I do find the original implementation as completely legit usage of plain standard library, e.g. we are not adding unnecessary abstractions or strange stuffs ...
but not fussy on it done.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java (hunk 2 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java]: moved to plain `null` checks

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java (hunk 4 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java]: moved to plain `null` checks

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 10 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:77]: and is going to be much more generic:
https://github.com/java-operator-sdk/java-operator-sdk/pull/966

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 6 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:63]: ok, I'll do it tomorrow 👍

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 9 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:23]: Durability is already tested by other integration tests in this codebase.
But I can add a test on the reconciliation of the spec of the service, yes.

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 6 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java]: done

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 7 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java]: done

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 9 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:26]: added services durability integration tests

Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:26]: The test `testKeycloakCacheIsConnected` prove that Infinispan is connected.

**Comment by andreaTP** [comment]
@vmuzikar ready for next round! 🙂

File: .github/workflows/operator-ci.yml [hunk 11]
```diff
@@ -61,7 +61,7 @@ jobs:
       - name: Test operator running in cluster
         working-directory: operator
         run: |
-          eval $(minikube -p minikube docker-env) 
+          eval $(minikube -p minikube docker-env)
```
Comment (abstractj) [inline .github/workflows/operator-ci.yml:64]: @andreaTP no big deal, but I'm curious. Is this change necessary?

Comment (andreaTP) [reply .github/workflows/operator-ci.yml:64]: no, it's not strictly necessary, just good practice to remove trailing white spaces.

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+                .ignoreExceptions()
+                .untilAsserted(() -> {
+                    Log.info("Checking realm import has finished.");
+                    CRAssert.assertKeycloakRealmImportStatusCondition(realmImportSelector.get(), KeycloakRealmImportStatusCondition.DONE, true);
+                    Log.info("Checking Keycloak is stable.");
+                    CRAssert.assertKeycloakStatusCondition(crSelector.get(), KeycloakStatusCondition.READY, true);
+                });
+
+        Log.info("Testing the Keycloak Cluster");
+        Awaitility.await().atMost(5, MINUTES).ignoreExceptions().untilAsserted(() -> {
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:122]: I wonder, do we really need Awatility here? We know that KC is ready at this point, why should it fail? And i don't want to hear "just in case". 😉 That way we could set some global re-run policy for the whole testsuite.

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+                    .withLabels(Constants.DEFAULT_LABELS)
+                    .list()
+                    .getItems();
+
+            String token = null;
+            // Obtaining the token from the first pod
+            // Connecting using port-forward and a fixed port to respect the instance issuer used hostname
+            for (var pod: pods) {
+                Log.info("Testing Pod: " + pod.getMetadata().getName());
+                try (var portForward = k8sclient
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:136]: Please add some comment here for future generations with reasoning why we took this approach, why can't we use in-cluster curl.

File: operator/src/test/java/org/keycloak/operator/KeycloakServicesE2EIT.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.keycloak.operator.Constants.DEFAULT_LABELS;
+import static org.keycloak.operator.utils.K8sUtils.*;
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/KeycloakServicesE2EIT.java:23]: Please don’t use star imports.

File: operator/src/test/resources/token-test-realm.yaml [hunk 15]
```diff
@@ -0,0 +1,1725 @@
+apiVersion: keycloak.org/v2alpha1
```
Comment (vmuzikar) [inline operator/src/test/resources/token-test-realm.yaml:1]: Can't we re-use the example realm instead?

**Review by vmuzikar — CHANGES_REQUESTED** [review]

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        this.existingService = fetchExistingService();
+    }
+
+    private ServiceSpec getServiceSpec() {
+      return new ServiceSpecBuilder()
+              .addNewPort()
+              .withPort(Constants.KEYCLOAK_SERVICE_PORT)
+              .withProtocol(Constants.KEYCLOAK_SERVICE_PROTOCOL)
+              .endPort()
+              .withSelector(Constants.DEFAULT_LABELS)
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:49]: Two answers:

 - re: supporting multiple instances in the same namespace
 
I think that we should have this as a separate task, review all the resources that are created, and have proper testing in place to not break it by accident.

- re: this specific case

What would you suggest? Should we add a more specific label to the main deployment and propagate it to the Service? Or we can do it as a follow up for supporting multiple instances in one namespace (as this PR is already big enough)?

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java [hunk 17]
```diff
@@ -180,8 +180,8 @@ protected void savePodLogs() {
   }
 
   private static void setDefaultAwaitilityTimings() {
-    Awaitility.setDefaultPollInterval(Duration.ofSeconds(1));
-    Awaitility.setDefaultTimeout(Duration.ofSeconds(180));
+    Awaitility.setDefaultPollInterval(Duration.ofSeconds(2));
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:183]: not really relevant here, but reverted to 1

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+                .resources(Keycloak.class)
+                .inNamespace(kc.getMetadata().getNamespace())
+                .withName(kc.getMetadata().getName());
+        K8sUtils.deployKeycloak(k8sclient, kc, true);
+
+        var kcPodsSelector = k8sclient.pods().inNamespace(namespace).withLabel("app", "keycloak");
+
+        Keycloak keycloak = k8sclient.resources(Keycloak.class)
+                .inNamespace(namespace)
+                .list().getItems().get(0);
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:42]: agree, done 👍

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 12 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:122]: Because caches propagation is, at best of my understanding/experience, only eventually consistent.

I would be more than happy to remove `Awaitility`, let's give it a try in CI.

Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java]: This CI will give us some data:
https://github.com/andreaTP/keycloak/runs/5396139803?check_suite_focus=true

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+                    assertThat(username).isEqualTo("test");
+                }
+            }
+        });
+
+        // This is to test passing through the "Service", not 100% deterministic, but a smoke test that things are working as expected
+        // Executed here to avoid paying the setup time again
+        var service = new KeycloakService(k8sclient, kc);
+        Awaitility.await().atMost(5, MINUTES).ignoreExceptions().untilAsserted(() -> {
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java]: As above let's run a few tests in CI and evaluate if we can remove.

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 13 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:135]: The comment at line 133 was intended for this purpose, do you mind to suggest a different wording in case?

File: operator/src/test/java/org/keycloak/operator/KeycloakServicesE2EIT.java (hunk 14 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/KeycloakServicesE2EIT.java:23]: ok, done

File: operator/src/test/resources/token-test-realm.yaml (hunk 15 above)
Comment (andreaTP) [reply operator/src/test/resources/token-test-realm.yaml:1]: Unfortunately not, we are relying on the configuration of this specific hand crafted realm to run the test.

Also this is increasing the coverage for `RealmImport` CRs 🙂

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 12 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java]: Ok, the answer is that the CI doesn't pass:
https://github.com/andreaTP/keycloak/runs/5396139803?check_suite_focus=true#step:9:15317

I'll keep `Awaitility`

Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:121]: 👍

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:170]: 👍

File: operator/src/test/resources/token-test-realm.yaml (hunk 15 above)
Comment (vmuzikar) [reply operator/src/test/resources/token-test-realm.yaml:1]: 👍

File: operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java (hunk 13 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java:135]: Nah, ok. Looking at it again and it seems sufficient. 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:49]: Ok, deal. 👍  This is most probably a bigger task, let's address it separately in #10562.

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks @andreaTP.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakService.java:49]: 👍

**Review by DGuhr — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Comment by andreaTP** [comment]
@abstractj the changes in this PR are completely self-contained in the `operator` module.
Any failure other than in `Keycloak Operator CI` job is not related.

More interestingly 🙂  :
A PR that compromise the clustering functionality of Keycloak can break the `Keycloak Operator CI` job but changes in the `operator` module cannot break any of the `Keycloak CI` jobs (at least so far 😇  ).

**Review by pedroigor — APPROVED** [review]

**Comment by abstractj** [comment]
@andreaTP thanks for clarifying Andrea. Not sure if it's related, but the Operator CI is broken in the main branch after the merge https://github.com/keycloak/keycloak/runs/5413342790?check_suite_focus=true

**Comment by andreaTP** [comment]
@abstractj opened #10587 , so far, it looks the only failure, but I'll make sure to double check the stability of the Operator CI 👍
